Sources/TuistAutomation/Simulator/SimulatorController.swift

```diff
@@ -0,0 +1,107 @@
+
+    /// Returns the list of simulator runtimes that are available in the system.
+    func runtimes() -> Single<[SimulatorRuntime]>
+
+    /// Returns the list of simulator devices and runtimes.
+    func devicesAndRuntimes() -> Single<[SimulatorDeviceAndRuntime]>
+}
+
+enum SimulatorControllerError: FatalError {
+    case simctlError(String)
+
+    var type: ErrorType {
+        switch self {
+        case .simctlError: return .abort
+        }
+    }
+
+    var description: String {
+        switch self {
+        case let .simctlError(output): return output
+        }
+    }
+}
+
+final class SimulatorController: SimulatorControlling {
+    private let jsonDecoder: JSONDecoder = JSONDecoder()
+
+    func devices() -> Single<[SimulatorDevice]> {
+        System.shared.observable(["/usr/bin/xcrun", "simctl", "list", "devices", "--json"])
+            .mapToString()
+            .collectOutput()
+            .asSingle()
+            .flatMap { output in
+                guard output.standardError.isEmpty else {
+                    return .error(SimulatorControllerError.simctlError(output.standardError))
+                }
+                do {
+                    let data = output.standardOutput.data(using: .utf8)!
+                    let json = try JSONSerialization.jsonObject(with: data, options: [])
+                    guard let dictionary = json as? [String: Any],
+                        let devicesJSON = dictionary["devices"] as? [String: [[String: Any]]] else {
+                        return .just([])
+                    }
+
+                    let devices = try devicesJSON.flatMap { (runtimeIdentifier, devicesJSON) -> [SimulatorDevice] in
+                        try devicesJSON.map { (deviceJSON) -> SimulatorDevice in
+                            var deviceJSON = deviceJSON
+                            deviceJSON["runtimeIdentifier"] = runtimeIdentifier
+                            let deviceJSONData = try JSONSerialization.data(withJSONObject: deviceJSON, options: [])
+                            return try self.jsonDecoder.decode(SimulatorDevice.self, from: deviceJSONData)
+                        }
+                    }
+
+                    return .just(devices)
+                } catch {
+                    return .error(error)
+                }
+            }
+    }
+
+    func runtimes() -> Single<[SimulatorRuntime]> {
+        System.shared.observable(["/usr/bin/xcrun", "simctl", "list", "runtimes", "--json"])
+            .mapToString()
+            .collectOutput()
+            .asSingle()
+            .flatMap { output in
+                guard output.standardError.isEmpty else {
+                    return .error(SimulatorControllerError.simctlError(output.standardError))
+                }
+                do {
+                    let data = output.standardOutput.data(using: .utf8)!
+                    let json = try JSONSerialization.jsonObject(with: data, options: [])
```

Comment:
is there a reason we can't map directly to a codable object?

Reply:
Similar comment here. In this case the structure is slightly simpler:

```
{
  "runtimes": [
    {
      // runtime
    }
  ]
}
```

🤷‍♂️ I wonder why Apple didn't just return an array.